### ISSUE TITLE
CustomChange transactional behaviour different since 4.28.0

### ISSUE THREAD

**Issue by kristofdepypere** [liquibase/liquibase#5947]
### Search first

- [X] I searched and no similar issues were found

### Description

We are using PostgreSQL and are a long user of liquibase. Since the latest release `4.28.0` we observed an issue which we can only fix by rolling back to `4.27`.

With our existing ChangeSets that start our db state from scratch we now get the following error:

`Caused by: org.postgresql.util.PSQLException: ERROR: cannot ALTER TABLE "users" because it is being used by active queries in this session`

This happens on multiple places if we comment out certain problematic lines. It appears transactionality behaviour is different ? We pinpointed it is always related to a customChange. For example the following:

    <changeSet id="someid" author="myself
        <customChange class="<class implementing CustomTaskChange"></customChange>
        <addUniqueConstraint tableName="users" columnNames="mail_id, home_id"/>
    </changeSet>

In case the customChangeset uses the same table we get the error. It seems that previously the connection was committed/closed right after the customtaskchange and now the same one is used leading to the error we are getting.

We might fix this by revisting all the places we have a problem, however what is the correct and expected behaviour here, as this clearly changed since the latest version.

### Steps To Reproduce

See example in description

### Expected/Desired Behavior

The behaviour is as-is with previous versions.

### Liquibase Version

4.28.0

### Database Vendor & Version

PostgreSQL 14

### Liquibase Integration

gradle, spring boot, java

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

**Comment by kevin-atx** [comment]
@tati-qalified  - could you get more details and validate this issue?

**Comment by tati-qalified** [comment]
Hi @kristofdepypere, thank you for reporting this issue.
Would you be able to provide a changelog that's failing and the custom change you're using? Or directly a zip of the project if it's easier. That way we can better reproduce the bug and our developer team can point you to the part of the code where this can be fixed.

Thank you,
Tatiana

**Comment by lbadam** [comment]
@kristofdepypere I wanted to follow-up and see if you could provide the items requested above by @tati-qalified? Thanks!

**Comment by kristofdepypere** [comment]
Hi,

I trimmed it down to a small test case that clearly fails on `4.18.0`, but not on the version before.

[testcase.zip](https://github.com/user-attachments/files/16311363/testcase.zip)

When cleaning out one of our problematic `CustomTaskChange` classes, it turned out that it is related to using a `RowMapper` from Spring JDBC. It appears that the connection is not closed in `4.18.0` as it used to be. I tested this test case and verified that it fails on `4.18.0` and doesn't fail on a version before. 

You'll need spring JDBC in order to reproduce the issue.

**Comment by MalloD12** [comment]
Hi @kristofdepypere,

I've been looking into this issue but couldn't find yet what is causing it. For some reason when performing the CustomTaskChange execution (in `CustomChangeWrapper` class), it's holding the DB connection, what I was thinking is maybe we can set `autoCommit=true` before the custom task execution and then `autoCommit=false`. 

One other thing I noticed from your provided example is that each changeset (from the `changelog-master.xml` file) contains multiple changeObjects, and this is not a recommended use of it. For example:

```xml
<changeSet id="customChangeExecution" author="someone">
        <customChange class="com.example.demo.TestQuery" />
        <addUniqueConstraint schemaName="test" tableName="some_table" columnNames="id, name"/>
 </changeSet>
```

but if instead we have something like:

```xml
<changeSet id="customChangeExecution" author="someone">
        <customChange class="com.example.demo.TestQuery" />      
 </changeSet>
```
and 
```xml
<changeSet id="addUniqueConstraintExecution" author="someone">
        <addUniqueConstraint schemaName="test" tableName="some_table" columnNames="id, name"/>
 </changeSet>
```

this issue does not occur. These are my observations so far.

Thanks,
Daniel.

**Comment by Floppy012** [comment]
I'm not quite sure if this is related. But I observed that from 4.28 onwards liquibase will not write the creation of new columns to the changelog. We're using PostgreSQL as well.

I have added a new `@Column(name = "test")` to an existing model. Using the maven plugin's `liquibase:diff` goal in 4.28 or higher, will not result in any changelog file being created even though the output of the command shows that liquibase found the column:

```
[INFO] Found column test bigint
```

Downgrading to 4.27 and running the same command  will result in a correct changelog being generated:

```xml
<changeSet author="..." id="1722372582790-1">
    <addColumn tableName="my_table">
        <column name="test" type="bigint"/>
    </addColumn>
</changeSet>
```

**Comment by MalloD12** [comment]
Hi @Floppy012,

That doesn't sound related to the issue we were discussing here with Kristof, I would recommend you raise a separate issue for the behavior you are describing in your comment. 

Thanks,
Daniel.

**Comment by kristofdepypere** [comment]
We have a best practice to put logical changes together in the same `changeset` as this determines the transactional boundary. In some cases, this means triggering multiple changes in different ways. For example, first you add a column and in another you move data into that new column, and then you delete the old data.  You want to treat that as a unit of work.
Indeed, splitting these up will fix the issue, but it's not something we would like to do as we have years of db changes to go over and adjust, this will be a dangerous change. Question is why did the behavior changed.

**Comment by kristofdepypere** [comment]
Hi, 

Is there any progress on this ticket ? Or can we expect any progress soon ?

**Comment by kristofdepypere** [comment]
As we are stuck upgrading Liquibase since version 4.27.0, I spend a day on getting to the root cause. The change in behavior was in the end caused by a new Configuration setting introduced, enabled by default in 4.28.0: This is the related ticket DAT-17267.

When the ALWAYS_SET_FETCH_SIZE is enabled, the behavior is clearly different and sessions keep being active, blocking new queries accessing the same tables. At least in case of `CustomChange` changes. I have no idea why that is. Regardless, in order to unblock us, I managed to disable the ALWAYS_SET_FETCH_SIZE option using a Custom `ConfigurationValueProvider`, not very elegant, but it works.

We are unblocked here, still it does look like a problem that needs to be looked at closer as to why this can cause issues.

@MalloD12 @kevin-atx

**Comment by MalloD12** [comment]
Hi @kristofdepypere,

Thanks a lot for figuring out the root cause of this issue. I'll discuss it with the team, so we create the proper fix for this issue. Again, thank you for investing the time on finding the reason behind this.

Daniel.

**Comment by kristofdepypere** [comment]
@MalloD12 We did observe a big increase in times to run liquibase upgrades. A test suite that normally takes 45 minutes, now takes 1h 30 minutes. Is it too be expected it goes that much slower ? Is it because we needed to disable this `ALWAYS_SET_FETCH_SIZE` setting ? 
This suite is obviously doing a lot of upgrades, more than what we would do in production, but a normal test on a regular startup of our application also takes +_ 10 seconds longer. We are slightly concerned here.

**Comment by MalloD12** [comment]
Hi @kristofdepypere,

I'll check with other team members who are more experienced on this and let you know if there is anything we have recently merged that could be affecting here.  

Thanks,
Daniel.

**Comment by marcelstoer** [comment]
@kristofdepypere thank you so much for all the effort spent so far on this. We ran in exactly the same issue when upgrading from Spring Boot 3.3.5 which ships Liquibase 4.27.0 to Boot 3.4.0 that comes with Liquibase 4.29.2. At was at the early stages of debugging (2h) when I decided on a hunch to check the issues here.

Would you be willing to share your `ConfigurationValueProvider`?

Considering that you experience a massive increase in test execution time, however, we most likely will just downgrade Liquibase.

**Comment by kristofdepypere** [comment]
@marcelstoer here it is. Note that our testing is not done with realistic production scenario's and have a lot more change sets to be executed than a normal day scenario. Also, the 10 seconds additional startup time is when we execute our full Liquibase changelog, which is worth years of change sets in our case. Still, better don't have regressions in this area. 

In addition, we disabled the calling home analytics that is enabled by default now. We run our software at customers that would be very concerned if they would notice calls going out like this. 

```
public class CustomConfigurationValueProvider extends AbstractConfigurationValueProvider {

    private static final String ALWAYS_SET_FETCH_SIZE_KEY = SqlConfiguration.ALWAYS_SET_FETCH_SIZE.getKey();
    private static final String ANALYTICS_ENABLED_KEY = AnalyticsArgs.ENABLED.getKey();

    @Override
    public int getPrecedence() {
        return 600;
    }

    @Override
    public ProvidedValue getProvidedValue(String... keyAndAliases) {
        if (keyAndAliases == null || keyAndAliases.length == 0) {
            return null;
        }
        String currentKey = keyAndAliases[0];
        if (ALWAYS_SET_FETCH_SIZE_KEY.equals(currentKey)) {
            return new ProvidedValue(
                    ALWAYS_SET_FETCH_SIZE_KEY,
                    ALWAYS_SET_FETCH_SIZE_KEY,
                    "false",
                    "DB upgrade overridden Liquibase Configuration",
                    this);
        }
        if (ANALYTICS_ENABLED_KEY.equals(currentKey)) {
            return new ProvidedValue(
                    ANALYTICS_ENABLED_KEY,
                    ANALYTICS_ENABLED_KEY,
                    "false",
                    "DB upgrade overridden Liquibase Configuration",
                    this);
        }
        return null;
    }
}
```

**Comment by dpeger** [comment]
Any updates on this one? We've encountered the same problem with the update from spring-boot 3.3.7 to 3.4.1 (and thus the update from liquibase 4.27 to 4.29). As spring-boot 3.3.x will be EOL end of May 2025 I'd expect that more and more people will run into this problem. This problem is now around for almost 8 month and there doesn't seem to be any progress other than the workaround by @kristofdepypere to disable `liquibase.sql.alwaysSetFetchSize`.

I can confirm that this is caused by having a `customChange` together with another "normal" change like f.e. `addUniqueConstraint` or something similar in the same changeset. Our custom change is not doing very special things. It's just a few prepared statements to read and delete DB entries.

The suggestion to only have a single change per changeset is not really feasible. First of all, this issue is breaking changesets that are around for years. These can't easily be split up as liquibase will try to rexecute the new changesets. Yes one could add preconditions to skip the new steps, but this can become very cumbersome and complex. Second (as already mentioned by @kristofdepypere) grouping changes that logically (and also in terms of constraints) belong together in one changeset is the whole purpose of changesets.

I can however not confirm any performance regressions compared to 4.27. Actually database bootstrapping with 700+ changesets seems to take less time in a local setup (didn't test it thoroughly though).

**Comment by lbadam** [comment]
@dpeger thanks for following up on this. We have identified the root cause of this and will work to get the fix in an upcoming patch release. When we open a PR, we'll make sure it gets connected to this issue so that you - and others - can follow the progress.

**Comment by MalloD12** [comment]
Hi guys,

After a team discussion, we agreed to set only the fetch size for Oracle, which was the original database that made us implement the fetch size change. We can isolate this change to only apply to Oracle. We can do that in the next few days.

Thanks,
Daniel.
